---
title: "a-worked-example"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{a-worked-example}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  out.width = "100%"
)
```

```{r setup, warning=FALSE, message=FALSE}
library(FLORAL)
if (! "BiocManager" %in% installed.packages()) install.packages("BiocManager")
if (! "curatedMetagenomicData" %in% installed.packages()) BiocManager::install("curatedMetagenomicData")
if (! "patchwork" %in% installed.packages()) install.packages("patchwork")
library(curatedMetagenomicData)
library(dplyr)
library(patchwork)
```

# Using FLORAL for Microbiome Analysis

## Data
We will be using data from the [`curatedMetagenomicData`]() package. Take a look at the summary of the studies available:

```{r viewData}

curatedMetagenomicData::sampleMetadata  |> group_by(.data$study_name, .data$study_condition) |> count() |> arrange(.data$study_name) 

```

As an example, let us look at the `YachidaS_2019` study between healthy controls and colorectal cancer (CRC) patients.


```{r getData}
curatedMetagenomicData::curatedMetagenomicData("YachidaS_2019") 
# note --  if you are behind a firewall, see the solutions to 500 errors here:
# https://support.bioconductor.org/p/132864/
rawdata <- curatedMetagenomicData::curatedMetagenomicData("2021-03-31.YachidaS_2019.relative_abundance", dryrun = FALSE, counts = TRUE) |>  mergeData()
```

The authors here are more familiar with `phyloseq` objects; if you are a SummarizedExperiment guru, feel free to bypass this!

```{r transform, warning=FALSE, message=FALSE}
if (! "mia" %in% installed.packages()) BiocManager::install("mia")
crc_phy_raw <- mia::makePhyloseqFromTreeSummarizedExperiment(rawdata, abund_values = "relative_abundance")
```

## Running FLORAL

Here we extract the data from the `phyloseq` object to two matrices: the taxa matrix and the "outcomes" of whether a patient is healthy or has colorectal cancer (CRC). Note that for binary outcomes, the input vector `y` needs to be formatted with entries equal to either `0` or `1`. In addition, we need to specify `family = "binomial"` in `FLORAL` to fit the logistic regression model. To print the progress bar as the algorithm runs, please use `progress = TRUE`. 

```{r floral}
x <- phyloseq::otu_table(crc_phy_raw) |> t()

y <- phyloseq::sample_data(crc_phy_raw)[,"disease"] %>% t()

x <- x[y %in% c("CRC","healthy"),]
x <- x[,colSums(x >= 100) >= nrow(x)*0.2] # filter low abundance taxa

y <- as.numeric(as.factor(y[y %in% c("CRC","healthy")]))-1
fit <- FLORAL(x = x, y = y, family="binomial", ncv=10, progress=TRUE)
```

## Interpreting the Model

FLORAL, like other methods that have an optimization step, has two "best" solutions for $\lambda$ available: one minimizing the mean squared error ($\lambda_\min$), and one maximizing the value of $\lambda$ withing 1 standard error of the minimum mean squared error ($\lambda_{\text{1se}}$). These are referred to as the `min` and `1se` solutions, respectively.

We can see the mean squared error (MSE) and the coefficients vs log($\lambda$) as follows:

```{r plots,out.width = "50%",fig.height=4,fig.width=6,dpi=300}
fit$pmse + fit$pcoef
```

In both plots, the vertical dashed line and dotted line represent $\lambda_\min$ and $\lambda_{\text{1se}}$, respectively. In the MSE plot, the bands represent plus minus one standard error of the MSE. In the coefficient plot, the colored lines represent individual taxa, where taxa with non-zero values at $\lambda_\min$ and $\lambda_{\text{1se}}$ are selected as predictive of the outcome. 

To view specific names of the selected taxa, please see `fit$selected.feature$min` or `fit$selected.feature$1se` vectors. To view all coefficient estimates, please see `fit$best.beta$min` or `fit$best.beta$1se`. Without looking into ratios, one can crudely interpret positive or negative association between a taxon and the outcome by the positive or negative sign of the coefficient estimates. However, we recommend referring to the two-step procedure discussed below for a more rigorous interpretation based on ratios, which is derived from the log-ratio model assumption. 

```{r viewTaxa}

head(fit$selected.feature$min)

head(sort(fit$best.beta$min))

```

## The Two-step Procedure

In the previous section, we checked the lasso estimates without identifying specific ratios that are predictive of the outcome (CRC in this case). By default, `FLORAL` performs a two-step selection procedure to use `glmnet` and `step` regression to further identify taxa pairs which form predictive log-ratios. To view those pairs, use `fit$step2.ratios$min` or `fit$step2.ratios$1se` for names of ratios and `fit$step2.ratios$min.idx` or `fit$step2.ratios$1se.idx` for the pairs of indices in the original input count matrix `x`. Note that one taxon can occur in multiple ratios.

```{r view2step}

head(fit$step2.ratios$`1se`)

fit$step2.ratios$`1se.idx`

```

To further interpret the positive or negative associations between the outcome, please refer to the output `step` regression tables, where the effect sizes of the ratios can be found. 

While the corresponding p-values are also available, we recommend only using the p-values as a criterion to rank the strength of the association. We do not recommend directly reporting the p-values for inference, because these p-values were obtained after running the first step lasso model without rigorous post-selective inference. However, it is still valid to claim these selected log-ratios are predictive of the outcome, as demonstrated by the improved 10-fold cross-validated prediction errors.

```{r viewTable}

fit$step2.tables$`1se`

```