<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using FLORAL for Microbiome Analysis • FLORAL</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using FLORAL for Microbiome Analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">FLORAL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Using-FLORAL-for-Microbiome-Analysis.html">Using FLORAL for Microbiome Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Using-FLORAL-for-survival-models-with-longitudinal-microbiome-data.html">Using FLORAL for survival models with longitudinal microbiome data</a></li>
    <li><a class="dropdown-item" href="../articles/Using-FLORAL-with-phyloseq.html">Using FLORAL with phyloseq data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/vdblab/FLORAL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Using FLORAL for Microbiome Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/vdblab/FLORAL/blob/master/vignettes/Using-FLORAL-for-Microbiome-Analysis.Rmd" class="external-link"><code>vignettes/Using-FLORAL-for-Microbiome-Analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Using-FLORAL-for-Microbiome-Analysis.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vdblab.github.io/FLORAL/">FLORAL</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>We will be using data from the <code>curatedMetagenomicData</code>
package. For easier installation, we saved a flat copy of the data, but
the steps below show how that was created.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"BiocManager"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"curatedMetagenomicData"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"curatedMetagenomicData"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"patchwork"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"patchwork"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">curatedMetagenomicData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the summary of the studies available:</span></span>
<span><span class="fu">curatedMetagenomicData</span><span class="fu">::</span><span class="va">sampleMetadata</span>  <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">study_name</span>, <span class="va">.data</span><span class="op">$</span><span class="va">study_condition</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">study_name</span><span class="op">)</span> </span>
<span><span class="co">#As an example, let us look at the `YachidaS_2019` study between healthy controls and colorectal cancer (CRC) patients.</span></span>
<span><span class="fu">curatedMetagenomicData</span><span class="fu">::</span><span class="fu">curatedMetagenomicData</span><span class="op">(</span><span class="st">"YachidaS_2019"</span><span class="op">)</span> </span>
<span><span class="co"># note --  if you are behind a firewall, see the solutions to 500 errors here:</span></span>
<span><span class="co"># https://support.bioconductor.org/p/132864/</span></span>
<span><span class="va">rawdata</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="fu">::</span><span class="fu">curatedMetagenomicData</span><span class="op">(</span><span class="st">"2021-03-31.YachidaS_2019.relative_abundance"</span>, dryrun <span class="op">=</span> <span class="cn">FALSE</span>, counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="fu">mergeData</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu">assays</span><span class="op">(</span><span class="va">rawdata</span><span class="op">)</span><span class="op">$</span><span class="va">relative_abundance</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">rawdata</span><span class="op">@</span><span class="va">colData</span><span class="op">$</span><span class="va">disease</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span>,  file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"inst"</span>, <span class="st">"extdata"</span>, <span class="st">"YachidaS_2019.Rdata"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"YachidaS_2019.Rdata"</span>, package<span class="op">=</span><span class="st">"FLORAL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-floral">Running FLORAL<a class="anchor" aria-label="anchor" href="#running-floral"></a>
</h2>
<p>We extracted the data from the <code>TreeSummarizedExperiment</code>
object to two objects: the taxa matrix <code>x</code> and the “outcomes”
vector <code>y</code> of whether a patient is healthy or has colorectal
cancer (CRC). Note that for binary outcomes, the input vector
<code>y</code> needs to be formatted with entries equal to either
<code>0</code> or <code>1</code>. In addition, we need to specify
<code>family = "binomial"</code> in <code>FLORAL</code> to fit the
logistic regression model. To print the progress bar as the algorithm
runs, please use <code>progress = TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">y</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRC"</span>,<span class="st">"healthy"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">*</span><span class="fl">0.2</span><span class="op">]</span> <span class="co"># filter low abundance taxa</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,split<span class="op">=</span><span class="st">"[|]"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,split<span class="op">=</span><span class="st">"[|]"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">y</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRC"</span>,<span class="st">"healthy"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FLORAL.html">FLORAL</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, family<span class="op">=</span><span class="st">"binomial"</span>, ncv<span class="op">=</span><span class="fl">10</span>, progress<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using elastic net with a=1.Algorithm running for full dataset: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 1 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 2 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 3 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 4 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 5 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 6 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 7 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 8 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 9 out of 10: </span></span>
<span><span class="co">#&gt; Algorithm running for cv dataset 10 out of 10:</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interpreting-the-model">Interpreting the Model<a class="anchor" aria-label="anchor" href="#interpreting-the-model"></a>
</h2>
<p>FLORAL, like other methods that have an optimization step, has two
“best” solutions for
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>
available: one minimizing the mean squared error
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mo>min</mo></msub><annotation encoding="application/x-tex">\lambda_\min</annotation></semantics></math>),
and one maximizing the value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>
withing 1 standard error of the minimum mean squared error
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mtext mathvariant="normal">1se</mtext></msub><annotation encoding="application/x-tex">\lambda_{\text{1se}}</annotation></semantics></math>).
These are referred to as the <code>min</code> and <code>1se</code>
solutions, respectively.</p>
<p>We can see the mean squared error (MSE) and the coefficients vs
log(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>)
as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span><span class="op">$</span><span class="va">pmse</span> <span class="op">+</span> <span class="va">fit</span><span class="op">$</span><span class="va">pcoef</span></span></code></pre></div>
<p><img src="Using-FLORAL-for-Microbiome-Analysis_files/figure-html/plots-1.png" class="r-plt" alt="" width="100%"></p>
<p>In both plots, the vertical dashed line and dotted line represent
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mo>min</mo></msub><annotation encoding="application/x-tex">\lambda_\min</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mtext mathvariant="normal">1se</mtext></msub><annotation encoding="application/x-tex">\lambda_{\text{1se}}</annotation></semantics></math>,
respectively. In the MSE plot, the bands represent plus minus one
standard error of the MSE. In the coefficient plot, the colored lines
represent individual taxa, where taxa with non-zero values at
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mo>min</mo></msub><annotation encoding="application/x-tex">\lambda_\min</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mtext mathvariant="normal">1se</mtext></msub><annotation encoding="application/x-tex">\lambda_{\text{1se}}</annotation></semantics></math>
are selected as predictive of the outcome.</p>
<p>To view specific names of the selected taxa, please see
<code>fit$selected.feature$min</code> or
<code>fit$selected.feature$1se</code> vectors. To view all coefficient
estimates, please see <code>fit$best.beta$min</code> or
<code>fit$best.beta$1se</code>. Without looking into ratios, one can
crudely interpret positive or negative association between a taxon and
the outcome by the positive or negative sign of the coefficient
estimates. However, we recommend referring to the two-step procedure
discussed below for a more rigorous interpretation based on ratios,
which is derived from the log-ratio model assumption.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">selected.feature</span><span class="op">$</span><span class="va">min</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "s__Actinomyces_sp_HMSC035G02"      "s__Actinomyces_sp_ICM47"          </span></span>
<span><span class="co">#&gt; [3] "s__Anaerotignum_lactatifermentans" "s__Anaerotruncus_colihominis"     </span></span>
<span><span class="co">#&gt; [5] "s__Asaccharobacter_celatus"        "s__Bacteroides_caccae"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">best.beta</span><span class="op">$</span><span class="va">min</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;             s__Parvimonas_micra      s__Collinsella_aerofaciens </span></span>
<span><span class="co">#&gt;                     -0.07868135                     -0.04559018 </span></span>
<span><span class="co">#&gt;    s__Actinomyces_sp_HMSC035G02        s__Holdemania_filiformis </span></span>
<span><span class="co">#&gt;                     -0.04128899                     -0.03473013 </span></span>
<span><span class="co">#&gt;    s__Anaerotruncus_colihominis s__Bacteroides_cellulosilyticus </span></span>
<span><span class="co">#&gt;                     -0.02883898                     -0.02154445</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-two-step-procedure">The Two-step Procedure<a class="anchor" aria-label="anchor" href="#the-two-step-procedure"></a>
</h2>
<p>In the previous section, we checked the lasso estimates without
identifying specific ratios that are predictive of the outcome (CRC in
this case). By default, <code>FLORAL</code> performs a two-step
selection procedure to use <code>glmnet</code> and <code>step</code>
regression to further identify taxa pairs which form predictive
log-ratios. To view those pairs, use <code>fit$step2.ratios$min</code>
or <code>fit$step2.ratios$1se</code> for names of ratios and
<code>fit$step2.ratios$min.idx</code> or
<code>fit$step2.ratios$1se.idx</code> for the pairs of indices in the
original input count matrix <code>x</code>. Note that one taxon can
occur in multiple ratios.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">step2.ratios</span><span class="op">$</span><span class="va">`1se`</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "s__Bacteroides_plebeius/s__Roseburia_faecis"                    </span></span>
<span><span class="co">#&gt; [2] "s__Eubacterium_eligens/s__Bacteroides_cellulosilyticus"         </span></span>
<span><span class="co">#&gt; [3] "s__Collinsella_aerofaciens/s__Bifidobacterium_pseudocatenulatum"</span></span>
<span><span class="co">#&gt; [4] "s__Collinsella_aerofaciens/s__Coprobacter_fastidiosus"          </span></span>
<span><span class="co">#&gt; [5] "s__Streptococcus_salivarius/s__Parvimonas_micra"                </span></span>
<span><span class="co">#&gt; [6] "s__Holdemania_filiformis/s__Eubacterium_sp_CAG_38"</span></span>
<span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">step2.ratios</span><span class="op">$</span><span class="va">`1se.idx`</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></span>
<span><span class="co">#&gt; [1,]   NA    1    4   12   12   22   27   27   64    73</span></span>
<span><span class="co">#&gt; [2,]   NA   81  113  102  114  131   90  151  131    80</span></span></code></pre></div>
<p>To further interpret the positive or negative associations between
the outcome, please refer to the output <code>step</code> regression
tables, where the effect sizes of the ratios can be found.</p>
<p>While the corresponding p-values are also available, we recommend
only using the p-values as a criterion to rank the strength of the
association. We do not recommend directly reporting the p-values for
inference, because these p-values were obtained after running the first
step lasso model without rigorous post-selective inference. However, it
is still valid to claim these selected log-ratios are predictive of the
outcome, as demonstrated by the improved 10-fold cross-validated
prediction errors.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">step2.tables</span><span class="op">$</span><span class="va">`1se`</span></span>
<span><span class="co">#&gt;                                                                    Estimate</span></span>
<span><span class="co">#&gt; (Intercept)                                                     -0.27768687</span></span>
<span><span class="co">#&gt; s__Bacteroides_plebeius/s__Roseburia_faecis                     -0.03017163</span></span>
<span><span class="co">#&gt; s__Eubacterium_eligens/s__Bacteroides_cellulosilyticus           0.04757745</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Bifidobacterium_pseudocatenulatum -0.03155717</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Coprobacter_fastidiosus           -0.04228961</span></span>
<span><span class="co">#&gt; s__Streptococcus_salivarius/s__Parvimonas_micra                  0.07442538</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Eubacterium_sp_CAG_38               -0.04151069</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Proteobacteria_bacterium_CAG_139    -0.03977612</span></span>
<span><span class="co">#&gt; s__Clostridium_aldenense/s__Parvimonas_micra                     0.07865537</span></span>
<span><span class="co">#&gt; s__Lachnospira_pectinoschiza/s__Ruminococcus_torques             0.03308018</span></span>
<span><span class="co">#&gt;                                                                 Std. Error</span></span>
<span><span class="co">#&gt; (Intercept)                                                     0.21517691</span></span>
<span><span class="co">#&gt; s__Bacteroides_plebeius/s__Roseburia_faecis                     0.01108305</span></span>
<span><span class="co">#&gt; s__Eubacterium_eligens/s__Bacteroides_cellulosilyticus          0.01352613</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Bifidobacterium_pseudocatenulatum 0.01570154</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Coprobacter_fastidiosus           0.01670159</span></span>
<span><span class="co">#&gt; s__Streptococcus_salivarius/s__Parvimonas_micra                 0.02220180</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Eubacterium_sp_CAG_38               0.01573689</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Proteobacteria_bacterium_CAG_139    0.01768439</span></span>
<span><span class="co">#&gt; s__Clostridium_aldenense/s__Parvimonas_micra                    0.02322619</span></span>
<span><span class="co">#&gt; s__Lachnospira_pectinoschiza/s__Ruminococcus_torques            0.01260594</span></span>
<span><span class="co">#&gt;                                                                   z value</span></span>
<span><span class="co">#&gt; (Intercept)                                                     -1.290505</span></span>
<span><span class="co">#&gt; s__Bacteroides_plebeius/s__Roseburia_faecis                     -2.722323</span></span>
<span><span class="co">#&gt; s__Eubacterium_eligens/s__Bacteroides_cellulosilyticus           3.517448</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Bifidobacterium_pseudocatenulatum -2.009814</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Coprobacter_fastidiosus           -2.532071</span></span>
<span><span class="co">#&gt; s__Streptococcus_salivarius/s__Parvimonas_micra                  3.352223</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Eubacterium_sp_CAG_38               -2.637794</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Proteobacteria_bacterium_CAG_139    -2.249223</span></span>
<span><span class="co">#&gt; s__Clostridium_aldenense/s__Parvimonas_micra                     3.386494</span></span>
<span><span class="co">#&gt; s__Lachnospira_pectinoschiza/s__Ruminococcus_torques             2.624173</span></span>
<span><span class="co">#&gt;                                                                     Pr(&gt;|z|)</span></span>
<span><span class="co">#&gt; (Intercept)                                                     0.1968753868</span></span>
<span><span class="co">#&gt; s__Bacteroides_plebeius/s__Roseburia_faecis                     0.0064824806</span></span>
<span><span class="co">#&gt; s__Eubacterium_eligens/s__Bacteroides_cellulosilyticus          0.0004357183</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Bifidobacterium_pseudocatenulatum 0.0444509033</span></span>
<span><span class="co">#&gt; s__Collinsella_aerofaciens/s__Coprobacter_fastidiosus           0.0113391045</span></span>
<span><span class="co">#&gt; s__Streptococcus_salivarius/s__Parvimonas_micra                 0.0008016534</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Eubacterium_sp_CAG_38               0.0083447178</span></span>
<span><span class="co">#&gt; s__Holdemania_filiformis/s__Proteobacteria_bacterium_CAG_139    0.0244983246</span></span>
<span><span class="co">#&gt; s__Clostridium_aldenense/s__Parvimonas_micra                    0.0007079184</span></span>
<span><span class="co">#&gt; s__Lachnospira_pectinoschiza/s__Ruminococcus_torques            0.0086859550</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generating-taxa-selection-probabilities">Generating taxa selection probabilities<a class="anchor" aria-label="anchor" href="#generating-taxa-selection-probabilities"></a>
</h2>
<p>It is encouraged to run k-fold cross-validation for several times to
account for the random fold splits. <code>FLORAL</code> provides
<code>mcv.FLORAL</code> functions to repeat cross-validations for
<code>mcv</code> times and on <code>ncore</code> cores. The output
summarizes taxa selection probabilities, average coefficients based on
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mo>min</mo></msub><annotation encoding="application/x-tex">\lambda_\min</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>λ</mi><mtext mathvariant="normal">1se</mtext></msub><annotation encoding="application/x-tex">\lambda_{\text{1se}}</annotation></semantics></math>.
Interpretable plots can be created if <code>plot = TRUE</code> is
specified.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mcv.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mcv.FLORAL.html">mcv.FLORAL</a></span><span class="op">(</span>mcv<span class="op">=</span><span class="fl">2</span>,</span>
<span>                      ncore<span class="op">=</span><span class="fl">1</span>,</span>
<span>                      x <span class="op">=</span> <span class="va">x</span>, </span>
<span>                      y <span class="op">=</span> <span class="va">y</span>, </span>
<span>                      family <span class="op">=</span> <span class="st">"binomial"</span>, </span>
<span>                      ncv <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                      progress<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in mcv.FLORAL(mcv = 2, ncore = 1, x = x, y = y, family = "binomial", :</span></span>
<span><span class="co">#&gt; Using 1 core for computation.</span></span>
<span><span class="co">#&gt; Random 3-fold cross-validation: 1</span></span>
<span><span class="co">#&gt; Random 3-fold cross-validation: 2</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mcv.fit</span><span class="op">$</span><span class="va">p_min</span></span></code></pre></div>
<p><img src="Using-FLORAL-for-Microbiome-Analysis_files/figure-html/mcvplots-1.png" class="r-plt" alt="" width="100%"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#Other options are also available</span></span>
<span><span class="co">#mcv.fit$p_min_ratio</span></span>
<span><span class="co">#mcv.fit$p_1se</span></span>
<span><span class="co">#mcv.fit$p_1se_ratio</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="elastic-net">Elastic net<a class="anchor" aria-label="anchor" href="#elastic-net"></a>
</h2>
<p>Beyond lasso model, <code>FLORAL</code> also supports elastic net
models by specifying the tuning parameter <code>a</code> between 0 and
1. Lasso penalty will be used when <code>a=1</code> while ridge penalty
will be used when <code>a=0</code>.</p>
<p>The <code>a.FLORAL</code> function can help investigate the
prediction performance for different choices of <code>a</code> and
return a plot of the corresponding prediction metric trajectories
against the choice of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">a.fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/a.FLORAL.html">a.FLORAL</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                  x <span class="op">=</span> <span class="va">x</span>, </span>
<span>                  y <span class="op">=</span> <span class="va">y</span>, </span>
<span>                  family <span class="op">=</span> <span class="st">"binomial"</span>, </span>
<span>                  ncv <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                  progress<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in a.FLORAL(a = c(0.1, 1), ncore = 1, x = x, y = y, family =</span></span>
<span><span class="co">#&gt; "binomial", : Using 1 core for computation.</span></span>
<span><span class="co">#&gt; Running for a = 0.1</span></span>
<span><span class="co">#&gt; Running for a = 1</span></span>
<span></span>
<span><span class="va">a.fit</span></span></code></pre></div>
<p><img src="Using-FLORAL-for-Microbiome-Analysis_files/figure-html/a.floral-1.png" class="r-plt" alt="" width="50%"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Teng Fei, Tyler Funnell, Nicholas Waters, Sandeep Raj.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
